### PR TITLE
 Implemented track-wise upcoming sessions feature

### LINKED ISSUE #1785: Upcoming/ongoing sessions Trackwise feature
**Actual Behaviour**

There is no way to find out the upcoming and ongoing sessions in a particular track other than manually searching for it. 

**Expected Behaviour**

There should be an option that displays the ongoing/upcoming sessions by auto scrolling to the ongoing/upcoming session, including a label on every session card.

**Screenshots of the issue**

I made some mock-ups

device-2017-07-06-202347

device-2017-07-06-202505

**Would you like to work on the issue?**

Yes.

**Target Milestone to solve the issue**

Till which [milestone](https://github.com/fossasia/open-event-android/milestones) will you be fixing 
the issue? 
https://github.com/fossasia/open-event-android/milestone/32

----

@the-dagger @mananwason @niranjan94 @mariobehling Shall I go ahead with this?

### PR DESCRIPTION
**Pull request by mahikaw**
Fixes #1785 
Changes: 

Screenshots for the change: 
In the issue itself.

### CODE REVIEW

**Comment by mahikaw** [comment]
@the-dagger @mananwason @niranjan94 @iamareebjamal @rohanagarwal94 @Shailesh351 @shreyasnbhat @arp95 Please review.

File: android/app/src/main/java/org/fossasia/openevent/activities/LocationActivity.java [hunk 1]
```diff
@@ -181,6 +181,7 @@ public void setUpcomingSession() {
         }
     }
 
+
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/LocationActivity.java]: Refrain from unneeded file changes

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java [hunk 2]
```diff
@@ -69,9 +73,14 @@
 
     private ActionBar actionBar;
 
-    @BindView(R.id.toolbar) Toolbar toolbar;
-    @BindView(R.id.recyclerView) RecyclerView sessionsRecyclerView;
-    @BindView(R.id.txt_no_sessions) TextView noSessionsView;
+    int ongoingPosition, upcomingPosition, flag;
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: Use private

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java [hunk 3]
```diff
@@ -134,6 +143,33 @@ private void loadData() {
 
             mSessions.clear();
             mSessions.addAll(track.getSessions().sort("startsAt"));
+
+//            finding upcoming and ongoing sessions
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: Comment slashes at right place please

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java [hunk 4]
```diff
@@ -134,6 +143,33 @@ private void loadData() {
...
+            Date current;
+            int countUpcoming = 0;
+            int countOngoing = 0;
+            for (Session trackSession : mSessions) {
+                flag = 0;
+                Date start = new Date();
+                Date end = new Date();
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: These can be initialized to null

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java [hunk 5]
```diff
@@ -134,6 +143,33 @@ private void loadData() {
...
+                if (((start.before(current)) || (start.equals(current))) && ((end.after(current)) || (end.equals(current)))) {
+                    ongoingPosition = countUpcoming;
+                    break;
+                } else if (start.after(current)) {
+                    upcomingPosition = countOngoing;
+                    break;
+
+                } else flag += 1;
+                countUpcoming += 1;
+                countOngoing += 1;
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: If everything here is shifted inside the try block, the initialization above won't be needed

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java [hunk 6]
```diff
@@ -134,6 +143,33 @@ private void loadData() {
...
+                    e.printStackTrace();
+                }
+                current = new Date();
+                if (((start.before(current)) || (start.equals(current))) && ((end.after(current)) || (end.equals(current)))) {
+                    ongoingPosition = countUpcoming;
+                    break;
+                } else if (start.after(current)) {
+                    upcomingPosition = countOngoing;
+                    break;
+
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: Extra line

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java [hunk 7]
```diff
@@ -222,6 +258,16 @@ public boolean onOptionsItemSelected(MenuItem item) {
         switch (item.getItemId()) {
             case R.id.action_search_sessions:
                 return true;
+            case R.id.upcoming_sessions:
+                if (ongoingPosition != 0)
+                    gridLayoutManager.scrollToPositionWithOffset(ongoingPosition, 0);
+                else if (upcomingPosition != 0)
+                    gridLayoutManager.scrollToPositionWithOffset(upcomingPosition, 0);
+                else {
+                    if ((flag > 0)) {
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: Use `else if` instead of nesting if in else

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java [hunk 8]
```diff
@@ -222,6 +258,16 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+                        Toast.makeText(this,getResources().getString(R.string.no_upcoming_ongoing), Toast.LENGTH_SHORT).show();
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: Space after comma and you can use getString directly

File: android/app/src/main/java/org/fossasia/openevent/adapters/SessionsListAdapter.java [hunk 9]
```diff
@@ -135,6 +137,27 @@ public void onBindViewHolder(final SessionViewHolder holder, final int position)
...
+        holder.sessionStatus.setVisibility(View.GONE);
+        DateUtils curr_date = new DateUtils();
+        Date start = new Date();
+        Date end = new Date();
+        try {
+            start = curr_date.getDate(session.getStartsAt());
+            end = curr_date.getDate((session.getEndsAt()));
+        } catch (ParseException e) {
+            e.printStackTrace();
+        }
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/adapters/SessionsListAdapter.java]: Same thing about the try block as above

File: android/app/src/main/java/org/fossasia/openevent/adapters/SessionsListAdapter.java [hunk 10]
```diff
@@ -135,6 +137,27 @@ public void onBindViewHolder(final SessionViewHolder holder, final int position)
...
+        if(start.compareTo(current)>0) {
+
+            holder.sessionStatus.setVisibility(View.VISIBLE);
+            holder.sessionStatus.setText("UPCOMING");
+        }
+        else if((start.compareTo(current)<=0)&&(end.compareTo(current)>=0)){
+            holder.sessionStatus.setVisibility(View.VISIBLE);
+            holder.sessionStatus.setText("ONGOING");
+
+        }
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/adapters/SessionsListAdapter.java]: Delete extra lines and use date specific methods like before and after

File: android/app/src/main/res/layout/tracksactvity_item.xml [hunk 11]
```diff
@@ -105,14 +105,21 @@
...
                         android:textSize="@dimen/text_size_large"
                         tools:text="Speaker" />
-                </LinearLayout>
+                    <TextView
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:id="@+id/session_status"
+                        android:layout_alignParentRight="true"
+                        android:textStyle="bold" />
+                </RelativeLayout>
```
Comment (iamareebjamal) [inline android/app/src/main/res/layout/tracksactvity_item.xml]: Use a horizontal linear layout instead

**Review by iamareebjamal — CHANGES_REQUESTED** [review]

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java (hunk 4 above)
Comment (mahikaw) [reply android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: Um, does it make a difference if we do that. I'm not aware, could you tell that please?

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java (hunk 3 above)
Comment (mahikaw) [reply android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: Do you mean that they aren't required here or that I should place them somewhere else. Could you be a bit more specific please?

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java (hunk 5 above)
Comment (mahikaw) [reply android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: Which initialisations are you talking about? `flag`, `countUpcoming` and `countOngoing`? I think they are still required. The IDE gives an initialisation error otherwise.

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java (hunk 4 above)
Comment (iamareebjamal) [reply android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: Object instantiation is a costly operation and should be avoided at all costs

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java (hunk 5 above)
Comment (iamareebjamal) [reply android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: I'm talking about Date object initialisations

Comment (mahikaw) [reply android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: Yes, that too gives an error. Could you have a look please?
screen shot 2017-07-19 at 2 50 41 pm

Comment (mahikaw) [reply android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: Should I initialise them to `null` ?

File: android/app/src/main/res/layout/tracksactvity_item.xml (hunk 11 above)
Comment (mahikaw) [reply android/app/src/main/res/layout/tracksactvity_item.xml]: Could you also explain the reason behind doing that please. I don't get why we need it 😅 
**EDIT**: Never mind, I get it now. I'll update it..

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java (hunk 5 above)
Comment (iamareebjamal) [reply android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: If you move it inside the try block, you won't need to initialise it to null

**Comment by mahikaw** [comment]
@iamareebjamal I have updated. Please review the changes.

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java [hunk 12]
```diff
@@ -134,6 +143,31 @@ private void loadData() {
...
+            int countUpcoming = 0;
+            int countOngoing = 0;
+            for (Session trackSession : mSessions) {
+                flag = 0;
+
+                try {
+                    Date start = DateUtils.getDate(trackSession.getStartsAt());
+                    Date end = DateUtils.getDate((trackSession.getEndsAt()));
+                    current = new Date();
+                    if (((start.before(current)) || (start.equals(current))) && ((end.after(current)) || (end.equals(current)))) {
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: there are some redundant wrapping brackets

File: android/app/src/main/java/org/fossasia/openevent/adapters/SessionsListAdapter.java [hunk 13]
```diff
@@ -135,6 +137,25 @@ public void onBindViewHolder(final SessionViewHolder holder, final int position)
             holder.sessionSubtitle.setVisibility(View.VISIBLE);
             holder.sessionSubtitle.setText(sessionSubTitle);
         }
+        holder.sessionStatus.setVisibility(View.GONE);
+        DateUtils curr_date = new DateUtils();
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/adapters/SessionsListAdapter.java]: DateUtils has static methods, no need to instantiate it

File: android/app/src/main/java/org/fossasia/openevent/adapters/SessionsListAdapter.java [hunk 14]
```diff
@@ -135,6 +137,25 @@ public void onBindViewHolder(final SessionViewHolder holder, final int position)
...
+        Date start;
+        Date end;
+        try {
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/adapters/SessionsListAdapter.java]: Can be moved inside try block

File: android/app/src/main/java/org/fossasia/openevent/adapters/SessionsListAdapter.java [hunk 15]
```diff
@@ -135,6 +137,25 @@ public void onBindViewHolder(final SessionViewHolder holder, final int position)
...
+            start = curr_date.getDate(session.getStartsAt());
+            end = curr_date.getDate((session.getEndsAt()));
+            Date current = new Date();
+            if(start.after(current)) {
+                holder.sessionStatus.setVisibility(View.VISIBLE);
+                holder.sessionStatus.setText("UPCOMING");
+            }
+            else if(((start.before(current)) || (start.equals(current))) && ((end.after(current)) || (end.equals(current)))){
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/adapters/SessionsListAdapter.java]: Move else if in above line and if this condition is same as used above, create a function returning a boolean for it

**Review by iamareebjamal — CHANGES_REQUESTED** [review]

**Comment by mahikaw** [comment]
@iamareebjamal I have updated. Please have a look now.

File: android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java [hunk 16]
```diff
@@ -173,4 +173,11 @@ public static String getRelativeTimeFromTimestamp(String timeStamp) throws Parse
                 (timeCreatedDate.getTime()),
                 System.currentTimeMillis(), android.text.format.DateUtils.SECOND_IN_MILLIS);
     }
+
+    public static boolean isUpcomingSession(Date start, Date end, Date current) {
+        if (((start.before(current)) || (start.equals(current))) && ((end.after(current)) || (end.equals(current))))
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java]: Please check this. And there are still some unneeded brackets here

**Comment by mahikaw** [comment]
@iamareebjamal I have updated please review.
@rohanagarwal94 @Shailesh351 @shreyasnbhat @arp95  I am rebasing it once again. Please review the PR!

File: android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java [hunk 17]
```diff
@@ -173,4 +173,11 @@ public static String getRelativeTimeFromTimestamp(String timeStamp) throws Parse
                 (timeCreatedDate.getTime()),
                 System.currentTimeMillis(), android.text.format.DateUtils.SECOND_IN_MILLIS);
     }
+
+    public static boolean isUpcomingSession(Date start, Date end, Date current) {
+        if ((start.before(current) || start.equals(current)) && (end.after(current) || end.equals(current)))
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java]: This is still not resolved

**Comment by mahikaw** [comment]
@iamareebjamal Updated. Please review.

File: android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java [hunk 18]
```diff
@@ -173,4 +173,9 @@ public static String getRelativeTimeFromTimestamp(String timeStamp) throws Parse
                 (timeCreatedDate.getTime()),
                 System.currentTimeMillis(), android.text.format.DateUtils.SECOND_IN_MILLIS);
     }
+
+    public static boolean isUpcomingSession(Date start, Date end, Date current) {
+        return ((start.before(current) || start.equals(current)) && (end.after(current) || end.equals(current)))
```
Comment (Shailesh351) [inline android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java]: semicolon missing.

**Review by Shailesh351 — CHANGES_REQUESTED** [review]
screenshot-1500652161263

@mahikaw Tested your PR. It is showing two search icons which was not before. Please fix it

**Comment by mahikaw** [comment]
Have updated @Shailesh351 Please check now.

File: android/app/src/main/java/org/fossasia/openevent/adapters/SessionsListAdapter.java [hunk 19]
```diff
@@ -129,19 +131,37 @@ public void onBindViewHolder(final SessionViewHolder holder, final int position)
...
             holder.sessionSubtitle.setVisibility(View.GONE);
         } else {
             holder.sessionSubtitle.setVisibility(View.VISIBLE);
             holder.sessionSubtitle.setText(sessionSubTitle);
         }
+        holder.sessionStatus.setVisibility(View.GONE);
+
+        try {
+            Date start;
+            Date end;
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/adapters/SessionsListAdapter.java]: Why is declaration detached from assignment?

File: android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java [hunk 20]
```diff
@@ -173,4 +173,9 @@ public static String getRelativeTimeFromTimestamp(String timeStamp) throws Parse
                 (timeCreatedDate.getTime()),
                 System.currentTimeMillis(), android.text.format.DateUtils.SECOND_IN_MILLIS);
     }
+
+    public static boolean isUpcomingSession(Date start, Date end, Date current) {
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java]: This is not the job of date utils, probably create a DateService. DateUtils only manages ISO8601 -> formatted date and vice versa

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java [hunk 21]
```diff
@@ -134,6 +143,31 @@ private void loadData() {
 
             mSessions.clear();
             mSessions.addAll(track.getSessions().sort("startsAt"));
+
+            //finding upcoming and ongoing sessions
+            Date current;
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: This is not used outside the loop, so it can be moved inside?

**Review by iamareebjamal — CHANGES_REQUESTED** [review]

File: android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java (hunk 20 above)
Comment (mahikaw) [reply android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java]: @iamareebjamal  I think that DateUtils. java shouldn't be restricted just to conversions as it is technically the 'utility' class for dates. What do you think?

Comment (iamareebjamal) [reply android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java]: As a concept of separation of concerns, a class should do only thing and do it well. DateUtils has no context of session and should not contain any method specific to it. Its sole purpose is to handle conversion of ISO8601 format dates

Comment (mahikaw) [reply android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java]: @iamareebjamal Okay, I'll add a new class for this then. Also, I'll also change the name for this class to dateConversion so that it doesn't confuse new developers..

**Comment by mahikaw** [comment]
@iamareebjamal Please check now. I have updated.

File: android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java (hunk 20 above)
Comment (iamareebjamal) [reply android/app/src/main/java/org/fossasia/openevent/utils/DateUtils.java]: DateConverter will be a better choice

File: android/app/src/main/java/org/fossasia/openevent/activities/LocationActivity.java [hunk 22]
```diff
@@ -155,7 +155,7 @@ public void setUpcomingSession() {
         Date current = new Date();
         for (Session sess : sortedSessions) {
             try {
-                Date start = DateUtils.getDate(sess.getStartsAt());
+                Date start = DateConversion.getDate(sess.getStartsAt());
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/LocationActivity.java]: DateConverter would be more suitable. And it was the work of another PR. This issue is not related to this change. Please make only related changes to keep work trackable

**Review by iamareebjamal — CHANGES_REQUESTED** [review]

File: android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java [hunk 23]
```diff
@@ -134,6 +144,30 @@ private void loadData() {
 
             mSessions.clear();
             mSessions.addAll(track.getSessions().sort("startsAt"));
+
+            //finding upcoming and ongoing sessions
+            int countUpcoming = 0;
+            int countOngoing = 0;
+            for (Session trackSession : mSessions) {
+                flag = 0;
+                Date current;
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java]: Date current can be moved to the place of assignment

File: android/app/src/main/java/org/fossasia/openevent/utils/DateService.java [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+/**
+ * Created by mw on 22/07/17.
+ */
+
+public class DateService {
+    public static boolean isUpcomingSession(Date start, Date end, Date current) {
+
+        return ((start.before(current) || start.equals(current)) && (end.after(current) || end.equals(current)));
+
+    }
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/utils/DateService.java]: Remove extra lines please

**Review by iamareebjamal — CHANGES_REQUESTED** [review]

**Comment by mahikaw** [comment]
@iamareebjamal Updated.
@rohanagarwal94 @Shailesh351 @shreyasnbhat @arp95 Please review this PR.

**Review by iamareebjamal — APPROVED** [review]
